content/general/functions.textile

```diff
@@ -8,17 +8,176 @@ Reactor Functions allow your server-less functions to be invoked following chann
+
+The contents of the @body@ depends on the format of your messages, and whether you've used an envelope or not.
+
+* No Envelope - Without an envelope, the contents of @body@ will simply be the contents of the message that was sent on the Ably Channel. 
+
+* Envelope - If you use an envelope, the @body@ will contain a bunch of extra attributes, which are the following:
+- source := the source of the request, namely "@channel.message"
+- appId := the ID of your application with Ably
+- channel := name of the channel the Ably Message was published on
+- site := an internal site identifier indicating the data center from which the message was published
+- ruleId := an ID for the rule which has created this request
+- presence := an @Array@ of presence events
+
+Each member of @presence@ has the following fields:
+
+- id := unique ID assigned by Ably to this message
+- clientId := the client identifier for the member. Find out more about "identified clients":https://support.ably.io/solution/articles/3000038608-authenticated-and-identified-clients
+- connectionId := the public unique identifier for the publisher's connection. Find out more about "@connectionId@":/realtime/connection#id
+- timestamp := the time in milliseconds since the epoch when this message was received by Ably
+- action := the presence action, as an integer defined by the "PresenceMessage Action enum":https://www.ably.io/documentation/realtime/types#presence-action.
+
+h4. Decoding presence events
+
+Each member of @presence@ is a 'raw' presence event. Presence messages sent "over the realtime service":/realtime/channels-messages are automatically decoded into "@PresenceMessages@":/realtime/types#presence-message object by the Ably client library. With Reactor Function you need to to do this explicitly, using "@PresenceMessage.fromEncodedArray@":/realtime/presence#presence-from-encoded-array on the @presence@ array, or "@PresenceMessage.fromEncoded@":/realtime/presence#presence-from-encoded on an individual member of that array. This will transform them into an array of "@PresenceMessages@":/realtime/types#presence-message (or in the case of @fromEncoded@, an individual "@PresenceMessage@":/realtime/types#presence-message). This has several advantages, e.g.:
+
+* It will decode the (numerical) action into a "@Presence action@":/realtime/presence#presence-action string (such as "@enter@", "@update@", or "@leave@")
+* It will fully decode any @data@ (using the @encoding@) back into the same datatype that it was sent in (or an equivalent in each client library's language)
+* If you are using "encryption":/realtime/encryption, you can pass your encryption key to the method and it will decrypt the @data@ for you
+
+We recommend you do this for all presence messages you receive through Reactor Functions. For example (using ably-js):
+
+```[javascript]
+const presMsgs = Ably.Realtime.PresenceMessage.fromEncodedArray(presenceMsg.data.presence);
```

Comment:
This is wrong as far as I can tell.  What is `presenceMsg.data`?  Surely it's more intuitive as `request.body.items`?